Hide Python dependency in `ConsolidateBlocks`

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
The following commits "hide" the Python dependency in the `run_consolidate_blocks` function as well as the `get_matrix_from_instruction` function from the `quantum_info` crate.

### Details and comments

The reason why I'm referring to this as "hiding" the python dependency instead of removing it, is that this workaround just limits the places where a `py` token is needed to only when Python is in use.

For example, in the both methods, we usually don't need a Py token to obtain the matrix of a `StandardGate` from Rust, since the definition is already in rust-space. However we will need one for any `PyGate` (which mostly represent custom defined gates), and to do so we require the usage of the `Operator` class. Since the need for `py` is only to call the `Operator` class in the case of a `PyGate` and `PyGate` can never be created from standalone mode, we assume that python is already in use and we manually obtain access to the `GIL`.

The one part that I'm not very satisfied with is the way we are hiding the `py` from the case when a block has more than 2 qubits. Since you can technically provide the nodes from Rust directly, there might be a case where a user might provide a block with more than 2 qubits. A warning has been added in the docstring of the `run_consolidate_blocks` method. However, I have an idea for a workaround by creating a wrapper around the function that accepts the same arguments, except for the ``blocks`` and ``runs``, which should be public to Rust users and might be the one to be used for the C API.

Feel free to comment with your thoughts.

raynelfss committed:
Remove `py` token from `get_matrix_from_inst`.
- Use `Python::with_gil` in exclusive cases where a `PyGate` is needed. Which is impossible to obtain directly from Rust.
- Use `Python::with_gil` in exclusive cases where the block has more than qargs. A comment needs to be added to explain this behavior (we should not be able to add blocks with more than 2 qubits).

raynelfss committed:
Docs: Document rust-side `run_consolidate_blocks`.

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -65,11 +65,32 @@ fn is_supported(
+/// If blocks are provided in standalone mode (Rust or C only) the blocks must not be
+/// operations with more than 2 qubits. Any consolidation with more than 2 qubits will use
+/// the Python class `Operator` which is not yet available through standalone mode and
+/// may lead to a panic.
```
I don't think this is likely to be a real issue we don't work with >2q blocks with any code that exists in rust. We control the C API so we just don't expose a `blocks` argument to the interface there so there is no possibility.

There is a transpiler pass that can do it in Python but nothing in Rust. We don't have it in the preset pass managers so it's not likely to be possible. But if you're concerned about it this much why not just split the function into two a Python entrypoint and rust entrypoint where the public rust entrypoint doesn't expose the `blocks` or `runs` arguments. Those only exists for Python. Then there is a private inner function that gets called by both entrypoints (and you can call the rust entrypoint from the python entrypoint if `blocks` and `runs` are None)

raynelfss replied:
> However, I have an idea for a workaround by creating a wrapper around the function that accepts the same arguments, except for the blocks and runs, which should be public to Rust users and might be the one to be used for the C API.

Seems we agree on this one thing :rocket:.

raynelfss replied:
Split out in 70c1143

raynelfss committed:
Add: Wrapper function exposed to Rust.
- Add wrapper function that exposes the same method but doesn't allow for the user to provide blocks or runs. The previous docstring is now assigned to the rust native wrapper and the warning has been removed.
- Make the python exclusive version of the method private.

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -349,7 +352,42 @@ pub fn run_consolidate_blocks(
+    basis_gate_name: &str,
+    force_consolidate: bool,
+    target: Option<&Target>,
+    basis_gates: Option<HashSet<String>>,
```
If this is the rust entrypoint for running consolidate blocks this should take a target unconditionally and basis_gates as a set isn't a thing:

```suggestion
    target: &Target,
```

raynelfss replied:
I will add this to the follow-on PR.

raynelfss replied:
Although, there are cases where you can run the Pass alone without a `Target` and it would just choose a `TwoQubitBasisDecomposer` with a `CXGate` for a base. Are these not going to be applicable from Rust?

mtreinish replied:
I think having an optional target is fine, there is a use case for this in the preset pass managers too (during the init stage). I had forgotten about that when I left the comment yesterday. But we don't need the basis gates set it should only be a target or none.

raynelfss replied:
This was removed from the follow-up PR, so we should be good to go.

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -349,7 +352,42 @@ pub fn run_consolidate_blocks(
+/// * `basis_gates` - The basis gates from which the decomposer will choose.
+pub fn run_consolidate_blocks(
+    dag: &mut DAGCircuit,
+    decomposer: DecomposerType,
```
This is not something we should need as an input, it's something that can be detected very quickly from the `Target`. But I know you have a PR for this already.

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -349,7 +352,42 @@ pub fn run_consolidate_blocks(
+/// * `force_consolidate` - Decides whether to force all consolidations or not.
+/// * `target` - The target representing the backend for which the pass is consolidating.
+/// * `basis_gates` - The basis gates from which the decomposer will choose.
+pub fn run_consolidate_blocks(
```
I feel like this should be in the follow on PR or you should combine the two PRs into one. If you're going to add a rust interface for running this pass we should use the signature that makes sense for calling it from rust. Ideally we'd have tests for this too, but it's ok for right now because there is no logic in this function it just calls the inner function.

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -183,22 +183,25 @@ pub fn run_consolidate_blocks(
                     dag.remove_op_node(node);
                 }
             } else {
-                let matrix = array.as_array().to_owned();
```
Oh nice catch that we can drop the clone here.

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -349,7 +352,42 @@ pub fn run_consolidate_blocks(
+/// This is the main function of the `ConsolidateBlocks` transpiler pass
+/// which replaces uninterrupted sequences of gates acting on the same qubits
+/// into a Unitary node, which will consecutively be resynthesized into a
+/// more optimal subcircuit.
```
This isn't entirely accurate. I would maybe reword it to something like:

```suggestion
/// This is function is the Rust entry point for the `ConsolidateBlocks` transpiler pass
/// which replaces uninterrupted sequences of gates acting on the same pair of qubits
/// into a [`UnitaryGate`] representing the unitary of that two qubit block if it estimated to
/// to optimize the circuit. This [`UnitaryGate`] subsequently will be synthesized by the
/// unitary synthesis pass into a more optimal subcircuit to replace that block.
```

mtreinish reviewed (commented):
The content of the existing function look good to me. I like the changes there and would be happy merging if the PR was just that. The concerns I have are on the new function, the intent looks like that function interface looks like the intent is to be rust entry point but the interface isn't really conducive for use from Rust or what we need for #14760. I know there is a follow-on PR to this that addresses some of my concerns maybe just remove the new function for now and integrate it into the follow up PR?

raynelfss committed:
Update crates/transpiler/src/passes/consolidate_blocks.rs

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

raynelfss committed:
Merge branch 'main' into remove-py-consolidate

raynelfss committed:
Remove: Rust entry point

mtreinish reviewed (approved):
LGTM, thanks for the updates
